### PR TITLE
[MNG-7700] test some edge cases with leading zeroes

### PR DESCRIPTION
**Pull request by elharo**
Also clean up docs, generics, and comparisons

### CODE REVIEW

File: maven-artifact/src/test/java/org/apache/maven/artifact/versioning/ComparableVersionTest.java [hunk 1]
```diff
@@ -209,6 +208,26 @@ public void testVersionComparing() {
...
+    @Test
+    public void testGetCanonical() {
+        // MNG-7700
+        newComparable("0.x");
+        newComparable("0-x");
+        newComparable("0.rc");
+        newComparable("0-1");
+
+        ComparableVersion version = new ComparableVersion("0.x");
+        assertEquals("x", version.getCanonical());
```
Comment (michael-o) [inline maven-artifact/src/test/java/org/apache/maven/artifact/versioning/ComparableVersionTest.java:226]: I still don't understand why this should not be 0 with `x` as qualifier.

Comment (elharo) [reply maven-artifact/src/test/java/org/apache/maven/artifact/versioning/ComparableVersionTest.java:226]: We never define or promise anything about the canonical representation of a ComparableVersion. Indeed, this method probably shouldn't have been public. All we promise is that two of them compare according to spec, and as best I can tell they do. For now this is just a characterization test of existing behavior. I'm leaving MNG-7700 open in case folks feel we should change the canonical representation. However neither that issue nor this PR needs to block the release.

Comment (michael-o) [reply maven-artifact/src/test/java/org/apache/maven/artifact/versioning/ComparableVersionTest.java:226]: Depressing...

Comment (dmlloyd) [reply maven-artifact/src/test/java/org/apache/maven/artifact/versioning/ComparableVersionTest.java:226]: This is not really a useful test. You're just testing that the output is what the output is. If you do this instead, you'll see that the "canonical" version is not really a canonical version at all:

```java
ComparableVersion version = new ComparableVersion("0.x");
ComparableVersion canonical = new ComparableVersion(version.getCanonical());
assertEquals(canonical, version); // actually see if the *versions* are equal
```

Comment (elharo) [reply maven-artifact/src/test/java/org/apache/maven/artifact/versioning/ComparableVersionTest.java:226]: I'm testing that the input produces the output. That is useful. There's a case to be made that the current behavior is wrong, which can be discussed on the bug. If we decide it is and change it, then having this test here will make it clear what changed when.

The test you suggest is also good. However, it's (not so obviously) performed in line 220 by `newComparable("0.x")` which includes an assert to check that.

**Review by elharo — COMMENTED** [review]
I'd like to merge this so I can proceed on top of it to investigate MNG-7714 which will need some more tests in these files.

**Review by michael-o — APPROVED** [review]
Fixup and merge please.

**Comment by slachiewicz** [comment]
please remember to close associated jira

**Comment by elharo** [comment]
See above. That Issue should not be closed because of this PR. JIRA issues and PRs have a many-to-many relationship. A PR can address more than one JIRA issue and a JIRA issue often requires more than one PR to fix.
